gpcontrib/arenadata_toolkit/arenadata_toolkit.c

```diff
@@ -1,4 +1,141 @@
+#include <dirent.h>
+#include <sys/stat.h>
+
 #include "postgres.h"
 #include "fmgr.h"
+#include "funcapi.h"
+#include "miscadmin.h"
+#include "storage/fd.h"
+#include "nodes/execnodes.h"
+#include "cdb/cdbvars.h"
+#include "utils/builtins.h"
+#include "utils/relfilenodemap.h"
+#include "utils/timestamp.h"
 
 PG_MODULE_MAGIC;
+
+#define should_skip(filename) \
+	(pg_strncasecmp(filename, "pg", 2) == 0 \
+	 || pg_strncasecmp(filename, "t_", 2) == 0 \
+	 || pg_strncasecmp(filename + strlen(filename) - 4, "_fsm", 4) == 0 \
+	 || pg_strncasecmp(filename + strlen(filename) - 3, "_vm", 3) == 0 \
+	 || pg_strncasecmp(filename + strlen(filename) - 5, "_init", 5) == 0)
+
+typedef struct
+{
+	char     *datpath;
+	DIR      *dirdesc;
+	TupleDesc tupdesc;
+} user_fctx_data;
+
+PG_FUNCTION_INFO_V1(adb_get_relfilenodes);
+Datum adb_get_relfilenodes(PG_FUNCTION_ARGS)
+{
+	Oid              datoid = MyDatabaseId;
+	Oid              tablespace_oid = MyDatabaseTableSpace;
+
+	struct dirent   *direntry;
+	char             filename[MAXPGPATH * 2];
+	user_fctx_data  *fctx_data;
+	FuncCallContext *funcctx;
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		MemoryContext oldcontext;
+
+		funcctx = SRF_FIRSTCALL_INIT();
+		oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
+
+		fctx_data = palloc(sizeof(user_fctx_data));
+		fctx_data->datpath = GetDatabasePath(datoid, tablespace_oid);
+		fctx_data->dirdesc = AllocateDir(fctx_data->datpath);
+
+		if (get_call_result_type(fcinfo, NULL, &fctx_data->tupdesc)
+				!= TYPEFUNC_COMPOSITE)
+			elog(ERROR, "return type must be a row type");
+
+		if (!fctx_data->dirdesc)
+			ereport(ERROR,
+					(errcode_for_file_access(),
+					 errmsg("could not read directory \"%s\": %m",
+							fctx_data->datpath)));
+
+		funcctx->attinmeta = TupleDescGetAttInMetadata(fctx_data->tupdesc);
+		funcctx->user_fctx = fctx_data;
+
+		MemoryContextSwitchTo(oldcontext);
+	}
+
+	funcctx = SRF_PERCALL_SETUP();
+	fctx_data = (user_fctx_data *) funcctx->user_fctx;
+
+	while ((direntry = ReadDir(fctx_data->dirdesc, fctx_data->datpath)) != NULL)
+	{
+		struct stat fst;
+		Datum       values[10];
+		bool        nulls[10];
+		char       *relfilenode;
+		Oid         reloid;
+		Oid         relfilenode_oid;
+		HeapTuple   tuple;
+
+		CHECK_FOR_INTERRUPTS();
+
+		if (direntry->d_type == DT_DIR)
+			continue;
+		if (should_skip(direntry->d_name))
+			continue;
+
+		snprintf(filename, sizeof(filename), "%s/%s",
```

Comment:
Why not do as `pg_logdir_ls` [does](https://github.com/arenadata/gpdb/blob/766d8607065d50a443357c5331a36f461bb90377/contrib/adminpack/adminpack.c#L369)?
Then it will not be necessary to allocate extra `MAXPGPATH * 2` bytes on the stack for the `filename` variable on each call.